buildlogger/client.go

```diff
@@ -0,0 +1,360 @@
+	TaskID      string
+	Execution   int32
+	TestName    string
+	Trial       int32
+	ProcessName string
+	Format      LogFormat
+	Arguments   map[string]string
+	Mainline    bool
+
+	// Storage location type for this log.
+	Storage LogStorage
+
+	// Configure a local sender for "fallback" operations and to collect
+	// the location of the buildlogger output.
+	Local send.Sender
+
+	// The number max number of bytes to buffer before sending log data
+	// over rpc to Cedar.
+	MaxBufferSize int
+
+	// Turn checking for new lines in messages off. If this is set to true
+	// make sure log messages do not contain new lines, otherwise the logs
+	// will be stored incorrectly.
+	NewLineCheckOff bool
+
+	// The gRPC client connection. If nil, a new connection will be
+	// established with the gRPC connection configuration.
+	ClientConn *grpc.ClientConn
+
+	// Configuration for gRPC client connection.
+	RPCAddress string
+	Insecure   bool
+	CAFile     string
+	CertFile   string
+	KeyFile    string
+
+	logID    string
+	exitCode int32
+}
+
+func (opts *LoggerOptions) validate() error {
+	if err := opts.Format.validate(); err != nil {
+		return err
+	}
+	if err := opts.Storage.validate(); err != nil {
+		return err
+	}
+
+	if opts.ClientConn == nil {
+		if opts.RPCAddress == "" {
+			return errors.New("must specify a RPC address when a client connection is not provided")
+		}
+		if !opts.Insecure && (opts.CAFile == "" || opts.CertFile == "" || opts.KeyFile == "") {
+			return errors.New("must specify credential files when making a secure connection over RPC")
+		}
+	}
+
+	if opts.Local == nil {
+		opts.Local = send.MakeNative()
+	}
+
+	if opts.MaxBufferSize == 0 {
+		// TODO: figure out ideal default size
+		opts.MaxBufferSize = 4096
+	}
+
+	return nil
+}
+
+// SetExitCode sets the exit code variable.
+func (opts *LoggerOptions) SetExitCode(i int32) { opts.exitCode = i }
+
+// GetLogID returns the unique buildlogger log ID set after NewLogger is
+// called.
+func (opts *LoggerOptions) GetLogID() string {
+	return opts.logID
+}
+
+// NewLogger returns a grip Sender backed by Cedar Buildlogger with level
+// information set.
+func NewLogger(ctx context.Context, name string, l send.LevelInfo, opts *LoggerOptions) (send.Sender, error) {
+	b, err := MakeLogger(ctx, name, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem making new logger")
+	}
+
+	if err := b.SetLevel(l); err != nil {
+		return nil, errors.Wrap(err, "problem setting grip level")
+	}
+
+	return b, nil
+}
+
+// MakeLogger returns a grip Sender backed by Cedar Buildlogger.
+func MakeLogger(ctx context.Context, name string, opts *LoggerOptions) (send.Sender, error) {
+	ts := time.Now()
+
+	if err := opts.validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid cedar buildlogger options")
+	}
+
+	var conn *grpc.ClientConn
+	var err error
+	if opts.ClientConn == nil {
+		rpcOpts := []grpc.DialOption{
+			grpc.WithUnaryInterceptor(aviation.MakeRetryUnaryClientInterceptor(10)),
```

Comment:
what does the 10 do? should that be a constant?

Reply:
Sets the number of retries to 10, doesn't have to be but do we want to make it configurable?